### PR TITLE
feat(op): support varlen npu flash attention

### PR DESCRIPTION
**Pull request by SolenoidWGT**
## Motivation
支持`torch_npu`的 var flash attention.

## Modification

移除了在pack数据下，给q和k进行padding和unpadding下的操作

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/model/ops/attention.py [hunk 1]
```diff
@@ -472,11 +604,13 @@ def _qkv_without_cu_seqlens(self, qkv, softmax_scale=None, causal=None, key_padd
             return _torch_fixedlen_qkvpacked_attn(qkv, self.dropout, softmax_scale, causal, key_padding_mask)
 
     @forward.register(conditions=(str(QKVPackType.KVPACKED), str(CuSeqlenType.WithOut)))
-    def _q_kv_without_cu_seqlens(self, q, kv, softmax_scale=None, causal=None, key_padding_mask=None):
+    def _q_kv_without_cu_seqlens(
+        self, q, kv, softmax_scale=None, causal=None, key_padding_mask=None, use_flash_attn=True
```
Comment (mwiacx) [inline internlm/model/ops/attention.py]: 不用把use_flash_attn作为attention forward的参数，这意味着model需要感知底层是否为fa的算子，我们不希望为model的开发者引入复杂度，这个东西作为算子选择系统的自动逻辑或者配置的一部分就好。

File: internlm/core/parallel/comm/tensor.py [hunk 2]
```diff
@@ -341,7 +341,7 @@ def output_hook(self, module: Embedding1D, args: Any, output: Tuple[Any]) -> Tup
         """
         _emb_dim = 2  # [bsz, seqlen, emb_dim]
 
-        return gather_forward_split_backward(output, self._parallel_mode, dim=_emb_dim)
+        return gather_forward_split_backward(output, self._parallel_mode, dim=_emb_dim), DUMMY_HANDLE_CONST
```
Comment (mwiacx) [inline internlm/core/parallel/comm/tensor.py]: 这里为啥要加DUMMY_HANDLE_CONST哇，他是作为register_forward_hook来用的，下同

File: internlm/model/ops/attention.py [hunk 3]
```diff
@@ -78,18 +77,15 @@ def _nyi_attn(func_name, *args, **kwargs):  # pylint: disable=W0613
 
 
 def _flash_float32_compatibility_wrapper(input_idxs: Tuple, flash_func: Callable, *args, **kwargs):
-    if gpc.config.model.dtype is torch.float32:
```
Comment (mwiacx) [inline internlm/model/ops/attention.py:81]: 这里为啥要去掉if

File: internlm/model/ops/attention.py [hunk 4]
```diff
@@ -252,20 +288,59 @@ def _npu_fixedlen_qkvsplited_attn(
         next_tockens=0,  # If necessary, expose the interface
         keep_prob=1 - dropout_p,
         inner_precise=0,  # If necessary, expose the interface
-    )
+    )[0]
 
 
-def _npu_varlen_qkvpacked_attn(
-    qkv: torch.Tensor, cu_seqlens, max_seqlen, dropout_p, softmax_scale=None, causal=False  # pylint: disable=W0613
+def __npu_varlen_qkvsplited_attn(
```
Comment (mwiacx) [inline internlm/model/ops/attention.py]: __npu -> _npu 一个下划线

File: internlm/core/parallel/comm/tensor.py (hunk 2 above)
Comment (SolenoidWGT) [reply internlm/core/parallel/comm/tensor.py]: 已移除

File: internlm/model/ops/attention.py (hunk 3 above)
Comment (SolenoidWGT) [reply internlm/model/ops/attention.py:81]: 这里之前想的是不在这个文件里import gpc方便单测，这里的修改已经回退

File: internlm/model/ops/attention.py (hunk 4 above)
Comment (SolenoidWGT) [reply internlm/model/ops/attention.py]: done

File: internlm/model/ops/attention.py (hunk 1 above)
Comment (SolenoidWGT) [reply internlm/model/ops/attention.py]: 这里之前想的是不在这个文件里import gpc方便单测，这些修改已经回退

File: internlm/core/parallel/comm/isp.py [hunk 5]
```diff
@@ -774,7 +774,12 @@ def auto_wrap_distributed_attention(cls: nn.Module) -> Callable[[bool, Any, floa
     def _attetion_constructor(
         local_attn_cls: type, causal=False, softmax_scale=None, attention_dropout=0.0
     ) -> nn.Module:
-        if gpc.config.parallel["tensor"].get("mode", "mtp") != "isp":
+        try:
```
Comment (sunpengsdu) [inline internlm/core/parallel/comm/isp.py:777]: 为啥要这么改

**Review by sunpengsdu — APPROVED** [review]

**Pull request merged**